### DISCUSSION TITLE
Better Typescript Types

### DISCUSSION THREAD

**Discussion by FroggyPanda** [supabase/supabase#13364] [Feature Requests]
The current way to extract row types from supabases typescript types is like so:
```ts
type Member = Database['public']['Tables']['member']['Row'];
```
which, in my opinion, is not the best way to do it.  Whenever I want to make sure my function return type is the same type as my database, I need to make a new type in that file or have another file that holds all of my table types.  I am thinking of on top of generating the `Database` type, supabase also generates the row types, so there is no need to do that long type abstraction when, in my opinion, it should be as easy as:
```ts
import { Member } from 'schema.ts';
```

Edit 1: Edited title from 'Better Typesciprt Row Types' to 'Better Typescript Types' as I feel [Barry's reply](https://github.com/orgs/supabase/discussions/13364#discussioncomment-5687872) and others can help improve the types generated 

Edit 2: Going to do a shameless plug for [better-supabase-type](https://github.com/FroggyPanda/better-supabase-types) and [Barry's supabase-react-query-codegen](https://github.com/barrymichaeldoyle/supabase-react-query-codegen).  Hope people who would like to see supabase have better first-party types find these packages useful, and the supabase team gets inspiration from these projects.

**Comment by Lyonsstzo** [comment]
I also think advocating for a more streamlined and user-friendly experience when using Supabase's TypeScript types.

**Comment by barrymichaeldoyle** [comment]
I ran into the same issue when building my [supabase-react-query-codegen CLI tool](https://github.com/barrymichaeldoyle/supabase-react-query-codegen) and ended up generating new types to help with generating my own hooks.

While doing that I figured that this is a problem to projects beyond the scope of just React Query and thought I might make a standalone CLI tool to just generate types.

But if we had a consensus on what we're expecting as a feature here (and clearly this raised Feature Request is a cry for for this feature). Then maybe it's worth me just making a PR and extending the Supabase type generation functionality.

**Example**

Let's say I use the [Supabase Type Generation](https://supabase.com/docs/guides/api/rest/generating-types) and end up with this file:
```
export type Json =
  | string
  | number
  | boolean
  | null
  | { [key: string]: Json }
  | Json[];

export interface Database {
  public: {
    Tables: {
      todo_items: {
        Row: {
          created_at: string;
          description: string;
          id: string;
          name: string;
        };
        Insert: {
          created_at?: string;
          description: string;
          id?: string;
          name: string;
        };
        Update: {
          created_at?: string;
          description?: string;
          id?: string;
          name?: string;
        };
      };
    };
    Views: {
      [_ in never]: never;
    };
    Functions: {
      [_ in never]: never;
    };
    Enums: {
      [_ in never]: never;
    };
    CompositeTypes: {
      [_ in never]: never;
    };
  };
}
```

Do you agree that it would be better to have something like this instead?
```
export type TodoItem = {
  created_at: string;
  description: string;
  id: string;
  name: string;
}

export type InsertTodoItem = {
  created_at?: string;
  description: string;
  id?: string;
  name: string;
}

export type UpdateTodoItem = {
  created_at?: string;
  description?: string;
  id?: string;
  name?: string;
}

export type Json =
  | string
  | number
  | boolean
  | null
  | { [key: string]: Json }
  | Json[];

export interface Database {
  public: {
    Tables: {
      todo_items: {
        Row: TodoItem;
        Insert: InsertTodoItem;
        Update: UpdateTodoItem;
      };
    };
    Views: {
      [_ in never]: never;
    };
    Functions: {
      [_ in never]: never;
    };
    Enums: {
      [_ in never]: never;
    };
    CompositeTypes: {
      [_ in never]: never;
    };
  };
}
```

Is this something you were envisioning? Or would you change anything with this implementation? I would love to hear your thoughts on this and get to work on making a proof of concept and a open a PR on the Supabase repo to implement this.

**Reply by FroggyPanda** [reply]
That looks exactly like what I was thinking of!  I was actually planning of either making a new discussion (which is probably the right idea as it needs website changes too) or adding it here.  But if there was a way to put in your own Json types too would be amazing.  For example I have a column that I want to only have the type of: `{name: string; age: number;}`.  But ill manage with the manual input for now.

**Reply by barrymichaeldoyle** [reply]
I did have a look at the CLI codebase and I could really find where the Type Generation occurs, or at least I didn't know how it worked.  It is written in Golang, and I'm more of a JS/TS guy. In my supabase-react-query-codegen project I use a bunch of TS functionality to generate the types.

I could go make a standalone library to generate the types which might be worth while, where you just provide a config on run a script which will generate the Database interface file and modify it on the spot, this is probably the best bet to work around waiting for the Supabase guys to tackle this. I might tackle this over the coming week purely because I think I'd use in it my future supabase projects.

**Reply by barrymichaeldoyle** [reply]
Regarding your use case to add your own JSON types, I haven't played around with Supabase in my projects enough to come across a use-case there to come up with a proper solution there.

**Reply by FroggyPanda** [reply]
> I could go make a standalone library to generate the types which might be worth while, [...]

I actually would like to stop you there.  I was trying to make this tool for this past week but your tool was such a big help I was finally able to finish it!

[Better Supabase Types](https://github.com/FroggyPanda/better-supabase-types) does what I was initially doing with making types from the supabase generated database and puts them into a new file.

**Comment by eligeske** [comment]
Being able to generate the types OOTB is really helpful!  Greatful for the feature and the work the Supabase team is doing.  As I'm using it I do have some opinions on the typing outputs as well.  One thing that I came across that would be helpful is using `const enum MyEnum {}` instead of Unions to reduce the code needed to create CRUD forms.  Const Enums have been a hot topic in the past as they were not  easily compiled in all projects. A type output config might scratch more itches? 

Again thanks for all the cool stuff!

**Comment by barrymichaeldoyle** [comment]
Legendary! I’ll go star it when I get back to my PC :)

**Comment by Blackfaded** [comment]
My solution to that "problem" is the following helper types:

```ts
export type Row<T extends keyof Database['public']['Tables']> = Database['public']['Tables'][T]['Row'];
export type InsertDto<T extends keyof Database['public']['Tables']> = Database['public']['Tables'][T]['Insert'];
export type UpdateDto<T extends keyof Database['public']['Tables']> = Database['public']['Tables'][T]['Update'];
```

So you can extract the types as needed with
```ts
import { InsertDto, Row, UpdateDto } from '@src/interfaces/database';

export type Location = Row<'location'>;
export type LocationInsertDto = InsertDto<'location'>;
export type LocationUpdateDto = UpdateDto<'location'>;
```

**Comment by jmarbutt** [comment]
So what about when you have more complex queries that you want to pass the type around. Say I have a query like this:

```js
const { data, error } = await supabase
  .from('messages')
  .select(`
    content,
    from:sender_id(name),
    to:receiver_id(name)
  `)
```
So if wanted to reuse the type that is resulted from this, how would you do it? Because I want to have a component that consumes this type in NextJS

**Reply by jmarbutt** [reply]
I think where I am struggling with this is take the example like this from the docs:

```js
const query = supabase.from('movies').select(`id, title`)
const movies: DbResult<typeof query> = await query
```

If I want to define the query to be used in various parts of the app, I don't have an instance of `supabase` in those cases. 

For example in NextJS if I created a file like this:

```js

import { createServerComponentClient } from "@supabase/auth-helpers-nextjs"

export function GetMovies() {

  const supabase = createServerComponentClient<Database>({ cookies })

  const query = supabase.from('movies').select(`id, title`)
  const movies: DbResult<typeof query> = await query
  
  return movies;
}
```

Now in multiple components, I may want to use the type of `DbResult<typeof query>` how could I make that a type that could be set as a parameter of a component? 

One reason this gets complicated from what I am trying to do, I have 2 apps that reuse queries and components but I define my supabase client different for each one like this:

```js
export function createSupabaseClient(accountId: string) {

    const options = {
        global: {
            headers: { 'x-account-id': accountId },
        },
    }
    
    const supabase = createServerComponentClient<Database>({ cookies }, { options })
    return supabase;
}

export function createPortalSupabaseClient() {
    const supabase = createServerComponentClient<Database>({ cookies })    
    return supabase;
}
```

I also may have other requirements in the header for the portal version.

**Reply by jmarbutt** [reply]
The only way I have been able to figure this out which seems like a bad idea is defining my query outside my components in NextJS like this:

```js

  const supabase = createServerComponentClient<Database>({ cookies })

 export const query = supabase.from('movies').select(`id, title`)
  
  export default async function Page() {

var results = await query;

return <MyComponent results={results} />

}
  
  ```
  
  Then my component looks like this:
  
  ```js
  `use client` 
  import query from './page.tsx'
  
  export default function MyComponent(params : { results: DbResult<typeof query> }) {

return <>{JSON.stringify(results.data)}</>

}
```

This is just a simple example of what I am trying to do.

**Reply by codewriter73** [reply]
In my current project I pass around the return types of complicated queries quite a bit and I've settled of two different ways of accomplishing that.

1. You can define the expected return type of the query outside of the function

```tsx
type SomeQuery = Tables<'table_one'> & {
  joined_table: Tables<'table_two'>;
  another_joined_table: Tables<'table_three'>;
}
```

Then you define the return type on the function itself and typescript will throw an error if there some how happens to be a mismatch.

```tsx
const query = async (): SomeQuery[] => {
  const { data } = await supabase.from('table_one').select(`
    *,
     joined_table: table_two(*),
     another_joined_table: table_three(*)
  `);

  return data;
}
```

The typescript utility types Omit and Pick become useful when you're not using the "*" select all. This way allows you to have a more fine grain control of the types that you export and use in other places

2. Use utility types to infer the return type of the function. A lot more simple but you lose some of the control, this is the one I use the most

```tsx
const query = async () => {
  const { data } = await supabase.from('table_one').select(`
    *,
     joined_table: table_two(*),
     another_joined_table: table_three(*)
  `);

  return data;
}

type SomeQuery = Awaited<ReturnType<typeof query>>;
```

**Reply by jmarbutt** [reply]
This seems like the best approach from what I can have seen, specially if you want to share these between something like a next and react native project. Although it would be cool if you could define the queries in a way to make them reusable and do something like :



```js

const myQuery = new SupabaseQuery<Table<'table_one'>>(`
  *,
     joined_table: table_two(*),
     another_joined_table: table_three(*)
`);

await supabase.query(myQuery)
```
> [!NOTE]  
> This is just a proposed idea, not an actual solution for anyone else just joining the conversation.

**Reply by codewriter73** [reply]
I think you just described a perfect use case for a custom hook if you're using them within the same project. Sharing between projects you're best bet might be to create something, either a function or a view, on the backend side so there's a single origin of truth for that query

**Reply by ThatXliner** [reply]
> This seems like the best approach from what I can have seen, specially if you want to share these between something like a next and react native project. Although it would be cool if you could define the queries in a way to make them reusable and do something like :
> 
> ```js
> const myQuery = new SupabaseQuery<Table<'table_one'>>(`
>   *,
>      joined_table: table_two(*),
>      another_joined_table: table_three(*)
> `);
> 
> await supabase.query(myQuery)
> ```
> 
> Note
> 
> This is just a proposed idea, not an actual solution for anyone else just joining the conversation.

I need this. The current workaround of instantiating a client is too limited and clunky

**Reply by ThatXliner** [reply]
Ok so I just hacked together a little thing

```typescript
import { GetResult } from "@supabase/postgrest-js/src/select-query-parser";
import { Database } from "./supabase/types";  // the auto-generated types
import { GenericTable } from "@supabase/supabase-js/dist/module/lib/types";
import { GenericSchema } from "@supabase/postgrest-js/dist/module/types";

type SchemaKeyType = keyof Database;
type TableKeyType = keyof Database[keyof Database]["Tables"];
export type RawSelect<
	SchemaName extends SchemaKeyType,
	TableName extends TableKeyType,
	Query extends string = "*",
	Schema extends GenericSchema = Database[SchemaName],
	Table extends GenericTable = Schema["Tables"][TableName],
	Relationships = Table extends { Relationships: infer R } ? R : unknown,
> = GetResult<Schema, Table["Row"], TableName, Relationships, Query>;

export type SelectPublic<
	Table extends TableKeyType,
	Query extends string = "*",
> = RawSelect<"public", Table, Query>;
```

To use

```typescript
const client = await createClient();  // create a Supabase client
const result: SelectPublic<
	"tableName",
	"query",
> = await client
	.from("tableName")
	.select("query")
// or, for no code duplication
const QUERY = "column1"
const otherResult: SelectPublic<
	"tableName",
	typeof QUERY,
> = await client
	.from("tableName")
	.select(QUERY)
```

**Comment by marc-wilson** [comment]
I like how there are separate types generated for the scenarios (Insert, Update, Row). If a flag could be added to the cli command to generate the types such as `--separate-types` (or something) it would be great if the result would be something like

```
// entity.type.ts
export interface Entity { ... }

// create-entity.type.ts
export interface CreateEntity { ... }

// update-entity.type.ts
export interface UpdateEntity { ... }
```

If there could be more "codegen-like" flags (commonly found in open api spec [code generators](https://github.com/OpenAPITools/openapi-generator/blob/master/docs/generators/typescript.md)) such as model prefix/suffix, it would be great too.

**Reply by Auguste-Houndjago** [reply]
Hi , 
I built a small utility that makes it easier to **write type-safe Supabase queries** with **automatic relation detection** and **full TypeScript autocompletion**

🔗 **GitHub:** [supabase-typed-query-builder](https://github.com/Auguste-Houndjago/supabase-typed-query-builder)  
📦 **NPM:** [supabase-typed-query-builder](https://www.npmjs.com/package/supabase-typed-query-builder)  
🎥 **Demo Video:**  
https://github.com/user-attachments/assets/43dae59e-0c2c-40be-bad2-a3ec96407b65

---

 Example usage

```ts
import { buildQuery } from "supabase-typed-query-builder"
import { supabase } from "@/utils/supabase/client"

const query = buildQuery("User", {
  select: {
    id: true,
    firstName: true,
    email: true,
  },
  include: {
    UserOrganization: {
      select: { role: true },
      include: {
        Organization: {
          select: { name: true }
        }
      }
    }
  }
})

const { data } = await supabase.from("User").select(query)
```

Works with @supabase-cache-helpers

```ts
"use client"
import { useQuery } from "@supabase-cache-helpers/postgrest-react-query"
import { supabase } from "@/utils/supabase/client"
import { buildQuery } from "supabase-typed-query-builder"

function getUser() {
  const query = buildQuery("User", {
    select: {
      id: true,
      firstName: true,
      lastName: true,
      email: true,
      avatar_url: true
    },
    include: {
      UserOrganization: {
        select: { role: true },
        include: {
          Organization: {
            select: { name: true, slug: true }
          }
        }
      }
    }
  })

  return supabase.from("User").select(query).eq("email", "augustehoundjago@gmail.com")
}

export default function Page() {
  const { data, error } = useQuery(getUser())

  return (
    <div style={{ padding: 16 }}>
      <h2>Get User</h2>
      <pre>{JSON.stringify(data, null, 2)}</pre>
      <pre>{JSON.stringify(error, null, 2)}</pre>
    </div>
  )
}

```

**Reply by StonedHackerman** [reply]
Awesome package! 

PS: I also noticed your package's NPM page has a broken link to your repo, probably bcs you renamed yourself on GitHub, you should fix it so people can find you easily
